tests/Benchmark/TextBenchmarkTests.mm

```diff
@@ -0,0 +1,587 @@
+                                                                        sc_defaultSize.height,
+                                                                        8,
+                                                                        sc_defaultSize.width * 4,
+                                                                        colorspace,
+                                                                        kCGImageAlphaPremultipliedFirst));
+    }
+
+protected:
+    woc::AutoCF<CGContextRef> m_context;
+};
+
+class CTFrameDrawBase : public TextBenchmarkBase {
+public:
+    CTFrameDrawBase() {
+        CTParagraphStyleSetting setting;
+        CTTextAlignment alignment = kCTCenterTextAlignment;
+        setting.spec = kCTParagraphStyleSpecifierAlignment;
+        setting.valueSize = sizeof(CTTextAlignment);
+        setting.value = &alignment;
+        auto paragraphStyle = woc::MakeAutoCF<CTParagraphStyleRef>(CTParagraphStyleCreate(&setting, std::extent<decltype(setting)>::value));
+        UIFont* font = [UIFont systemFontOfSize:24];
+
+        CFStringRef keys[2] = { kCTFontAttributeName, kCTParagraphStyleAttributeName };
+        CFTypeRef values[2] = { font, paragraphStyle };
+        auto dict = woc::MakeAutoCF<CFDictionaryRef>(CFDictionaryCreate(nullptr,
+                                                                        (const void**)keys,
+                                                                        (const void**)values,
+                                                                        std::extent<decltype(keys)>::value,
+                                                                        &kCFTypeDictionaryKeyCallBacks,
+                                                                        &kCFTypeDictionaryValueCallBacks));
+
+        auto attrString = woc::MakeAutoCF<CFAttributedStringRef>(CFAttributedStringCreate(nullptr, (CFStringRef)sc_frameText, dict));
+        auto framesetter = woc::MakeAutoCF<CTFramesetterRef>(CTFramesetterCreateWithAttributedString(attrString));
+        auto path = woc::MakeAutoCF<CGMutablePathRef>(CGPathCreateMutable());
+        CGPathAddRect(path, nullptr, CGRect{ CGPointZero, sc_defaultSize });
+        m_frame = woc::MakeAutoCF<CTFrameRef>(CTFramesetterCreateFrame(framesetter, CFRange{ 0, 0 }, path, nullptr));
+    }
+
+protected:
+    woc::AutoCF<CTFrameRef> m_frame;
+};
+
+class CTFrameDrawSingle : public CTFrameDrawBase {
+public:
+    inline void Run() {
+        CTFrameDraw(m_frame, m_context);
+    }
+
+    size_t GetRunCount() const {
+        return 1000;
+    }
+};
+
+class CTFrameDrawGroup : public CTFrameDrawBase {
+public:
+    void PreRun() {
+        _CGContextPushBeginDraw(m_context);
+    }
+
+    inline void Run() {
+        for (size_t i = 0; i < 100; ++i) {
+            CTFrameDraw(m_frame, m_context);
+        }
+    }
+
+    void PostRun() {
+        _CGContextPopEndDraw(m_context);
+    }
```

Comment:
If PostRun time doesn't count, you're ignoring how long it takes to _actually draw_ here; is that your intent?

With the new CG model, every frame draw will be batched and flushed once down here.